prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java

```diff
@@ -0,0 +1,267 @@
+                requestSigner,
+                Collections.emptyList());
+    }
+
+    public PrometheusSinkWriter(
+            ElementConverter<PrometheusTimeSeries, Types.TimeSeries> elementConverter,
+            Sink.InitContext context,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            int maxBatchSizeInSamples,
+            long maxTimeInBufferMS,
+            String prometheusRemoteWriteUrl,
+            CloseableHttpAsyncClient asyncHttpClient,
+            SinkCounters counters,
+            PrometheusRequestSigner requestSigner,
+            Collection<BufferedRequestState<Types.TimeSeries>> states) {
+        super(
+                elementConverter,
+                context,
+                maxBatchSizeInSamples, // maxBatchSize
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBatchSizeInSamples, // maxBatchSizeInBytes
+                maxTimeInBufferMS,
+                maxBatchSizeInSamples, // maxRecordSizeInBytes
+                states);
+        this.requestBuilder =
+                new PrometheusRemoteWriteHttpRequestBuilder(
+                        prometheusRemoteWriteUrl, requestSigner);
+        this.asyncHttpClient = asyncHttpClient;
+        this.counters = counters;
+    }
+
+    /**
+     * This is the "size" of the request entry (a {@link Types.TimeSeries time-series}) used for
+     * batching. Regardless the name of the method, it returns the number of {@link Types.Sample
+     * samples} in the time-series (not bytes), to support batching in terms of samples.
+     *
+     * @param timeSeries the request entry for which we want to know the size
+     * @return number of samples in the time-series
+     */
+    @Override
+    protected long getSizeInBytes(Types.TimeSeries timeSeries) {
+        return RequestEntrySizeUtils.requestSizeForBatching(timeSeries);
+    }
+
+    @Override
+    protected void submitRequestEntries(
+            List<Types.TimeSeries> requestEntries, Consumer<List<Types.TimeSeries>> requestResult) {
+        int timeSeriesCount = requestEntries.size();
+        long sampleCount = RequestEntrySizeUtils.countSamples(requestEntries);
+        if (LOG.isTraceEnabled()) {
+            LOG.trace(
+                    "Writing {} time-series containing {} samples ", timeSeriesCount, sampleCount);
+        }
+
+        Remote.WriteRequest writeRequest = buildWriteRequest(requestEntries);
+        byte[] requestBody;
+        try {
+            requestBody = compressWriteRequest(writeRequest);
+        } catch (IOException e) {
+            throw new RuntimeException("Exception compressing the request body", e);
+        }
+
+        SimpleHttpRequest postRequest = requestBuilder.buildHttpRequest(requestBody);
+        asyncHttpClient.execute(
+                postRequest,
+                new ResponseCallback(timeSeriesCount, sampleCount, counters, requestResult));
+    }
+
+    @VisibleForTesting
```

Comment:
Not a fan of this `@VisibleForTesting` - can we instead write a test  `asyncHttpClient` that tests each scenario?

Reply:
Moved the callback to a separate class and created a specific test.